sw/device/tests/flash_ctrl/flash_test.c

```diff
@@ -2,140 +2,155 @@
+                              /*erase_en=*/true);
+
+  // A memory protection region representing the first page of the second bank.
+  mp_region_t protection_region = {
+      .num = 0x0,
+      .base = FLASH_PAGES_PER_BANK,
+      .size = 0x1,
+      .rd_en = true,
+      .prog_en = true,
+      .erase_en = true,
+  };
+
+  uintptr_t ok_region_start =
+      FLASH_MEM_BASE_ADDR + (protection_region.base * FLASH_PAGE_SZ);
+  uintptr_t ok_region_end =
+      ok_region_start + (protection_region.size * FLASH_PAGE_SZ);
+  mmio_region_t ok_region = mmio_region_from_addr(ok_region_start);
+
+  uintptr_t bad_region_start = ok_region_end;
+
+  // Erase good and bad regions.
+  CHECK_EQZ(flash_page_erase(ok_region_start));
+  CHECK_EQZ(flash_page_erase(bad_region_start));
+
+  // Turn off flash access by default.
+  flash_default_region_access(/*rd_en=*/false, /*prog_en=*/false,
+                              /*erase_en=*/false);
+  flash_cfg_region(&protection_region);
+
+  // Attempt to perform a write.
+  uintptr_t region_boundary_start = bad_region_start - (FLASH_WORD_SZ * 3);
+  mmio_region_t region_boundary = mmio_region_from_addr(region_boundary_start);
+
+  uint32_t words[6];
+  memset(words, 0xa5, ARRAYSIZE(words) * sizeof(uint32_t));
+  for (int i = 0; i < ARRAYSIZE(words); ++i) {
+    words[i] += i;
   }
-  break_on_error(!flash_write(chk_addr, prog_array, good_words + bad_words));
 
-  // the good words should match
-  for (uint32_t i = 0; i < good_words; i++) {
-    if (REG32(chk_addr + i * 4) != prog_array[i]) {
-      break_on_error(1);
+  // Perform a partial write.
+  CHECK_NEZ(flash_write(region_boundary_start, words, ARRAYSIZE(words)));
+  // Words in the good region should still match, while words in the bad
+  // region should be all-ones, since we erased
+  for (int i = 0; i < ARRAYSIZE(words); ++i) {
+    uint32_t expected = 0xffffffff;
+    if (i < ARRAYSIZE(words) / 2) {
+      expected = words[i];
     }
+    CHECK(mmio_region_read32(region_boundary, i * sizeof(uint32_t)) ==
+          expected);
   }
 
-  // the bad word contents should not have gone through
-  for (uint32_t i = good_words; i < good_words + bad_words; i++) {
-    if (REG32(chk_addr + i * 4) != 0xFFFFFFFF) {
-      break_on_error(1);
-    }
+  // Attempt to erase bad page, which should fail.
+  CHECK_NEZ(flash_page_erase(bad_region_start));
+
+  // Attempt to erase the good page, which should succeed.
+  CHECK_EQZ(flash_page_erase(ok_region_start));
+  for (int i = 0; i < FLASH_WORDS_PER_PAGE; i++) {
+    CHECK_EQZ(~mmio_region_read32(ok_region, i * sizeof(uint32_t)));
   }
+}
 
-  // attempt to erase bad page, should error
-  break_on_error(!flash_page_erase(bad_addr_start));
+int main(int argc, char **argv) {
+  uart_init(UART_BAUD_RATE);
+  flash_init_block();
 
-  // erase the good page
-  break_on_error(flash_page_erase(good_addr_start));
+  flash_cfg_bank_erase(FLASH_BANK_0, /*erase_en=*/true);
+  flash_cfg_bank_erase(FLASH_BANK_1, /*erase_en=*/true);
```

Comment:
I strongly dislike inline comments in function calls.  Drop these here and then the `flash_ctrl` DIF can have proper types for them.

Reply:
I figure these will just get wiped out when the DIF gets written? For now I think we should keep them... it feels out of scope for this PR.